### PR TITLE
Support Callbacks in Pytorch Ray Estimators

### PR DESCRIPTION
**Pull request by charlieJ107**
Related Issue: https://github.com/intel-analytics/BigDL/issues/3625

- [x] add a `callbacks` argument in [estimator.fit](https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/orca/src/bigdl/orca/learn/pytorch/pytorch_ray_estimator.py#L196)

- [x] pass `callbacks` to [ray_runner.train_epochs](https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py#L239)

- [x] call each `callback`'s `set_model` and pass in the runner instance so that in each callback method the author can access runner through `self.model`

- [x] call each callback's on_train_begin

- [x] call each callback's on_epoch_begin before each epoch

- [x] pass callbacks to training_operator.train_epoch

- [x] call each callback's on_batch_begin before each batch

- [x] call each callback's on_batch_end after each batch

- [x] call each callbacks on_epoch_end after each epoch

- [x] call each callbacks on_train_end

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    def on_batch_end(self):
+        pass
+
+    def on_epoch_begin(self):
+        pass
+
+    def on_epoch_end(self):
+        pass
+
+    def on_predict_batch_begin(self):
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:17]: Let's only do training-related methods and remove `predict` and `test` related methods.

File: python/orca/src/bigdl/orca/learn/pytorch/pytorch_ray_estimator.py [hunk 2]
```diff
@@ -267,6 +269,11 @@ def transform_func(worker, partition_refs):
                                                        profile=profile,
                                                        info=info)
 
+        # TODO: Handle callbacks
+        for callback in callbacks:
+            from python.orca.src.bigdl.orca.learn.pytorch.callbacks import Callback
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/pytorch_ray_estimator.py:274]: only `bigdl.orca.learn.pytorch.callbacks`. src is the source root.

**Comment by jason-dai** [comment]
Why ray estimator only? Does it apply to pyspark estimator too?

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py (hunk 1 above)
Comment (charlieJ107) [reply python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:17]: Removed in next commmit.

**Comment by yangw1234** [comment]
> Why ray estimator only? Does it apply to pyspark estimator too?

It also works on pyspark estimator. We will add the `callbacks` arguments and corresponding changes in pyspark estimator in the following PR.

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 3]
```diff
@@ -124,6 +124,8 @@ def __init__(self,
         self.use_tqdm = use_tqdm
         self.scheduler_step_freq = scheduler_step_freq
         self.sync_stats = sync_stats
+        self.epochs_stats = None
+        self.epoch_stats = None
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:128]: maybe documents what's the differences of these two variables.

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 4]
```diff
@@ -251,9 +253,15 @@ def train_epochs(self, data_creator, epochs=1, batch_size=32, profile=False,
...
             loader = self.with_sampler(loader)
+
+        for callback in callbacks:
+            callback.set_model(self.models)
+            callback.on_train_begin(logs=self.epochs_stats)
         stats_list = list()
         for i in range(epochs):
-            stats = self.train_epoch(loader, profile=profile, info=info)
+            for callback in callbacks:
+                callback.on_epoch_begin(epoch=self.epochs, logs=stats_list[-1])
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:263]: what happens for the first epoch when `stats_list` is empty

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 5]
```diff
@@ -262,17 +270,26 @@ def train_epochs(self, data_creator, epochs=1, batch_size=32, profile=False,
...
+                    info=None,
+                    callbacks=None):
         """Runs a training epoch and updates the model parameters."""
         if hasattr(self.train_loader, "sampler") and hasattr(
                 self.train_loader.sampler, "set_epoch"):
             self.train_loader.sampler.set_epoch(self.epochs)
         self.logger.debug("Begin Training Step {}".format(self.epochs + 1))
+        for callback in callbacks:
+            # TODO: train_stats
+            callback.on_batch_begin(self.epochs, logs=self.epoch_stats)
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:292]: why calling `on_batch_begin` here? they should be called at the beginning of every batch, right?

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    def on_train_batch_end(self, batch, logs=None):
+        """
+        Called at the end of a training batch in `fit` methods.
+        Subclasses should override for any actions to run.
+        @param batch: Integer, index of batch within the current epoch.
+        @param logs:  Dict. Aggregated metric results up until this batch.
+        """
+        # For backwards compatibility.
+        self.on_batch_end(batch, logs=logs)
+        pass
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:72]: this `pass` is not needed right?

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    @abstractmethod
+    def on_train_batch_begin(self, batch, logs=None):
+        """
+        Called at the beginning of a training batch in `fit` methods.
+        Subclasses should override for any actions to run.
+        @param batch: Integer, index of batch within the current epoch.
+        @param logs: Dict. Currently, no data is passed to this argument for this method
+          but that may change in the future.
+        """
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:59]: call `on_batch_begin` here?

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+class Callback(object):
+    def __init__(self):
+        self.model = None
+        self.params = None
+
+    @abstractmethod
+    def on_batch_begin(self, batch, logs=None):
+        """
+        A backwards compatibility alias for `on_train_batch_begin`.
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:12]: which one do you expect user to override? I think only `on_batch_begin` and `on_batch_end` are enough and we can remove `on_train_batch_beigin` and `on_train_batch_end`

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+          the future.
+        """
+        pass
+
+    @abstractmethod
+    def set_model(self, model):
+        self.model = model
+
+    @abstractmethod
+    def set_param(self, param):
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:100]: where is method is called?

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        Called at the end of training.
+        Subclasses should override for any actions to run.
+        @param: Dict. Currently, the output of the last call to `on_epoch_end()`
+          is passed to this argument for this method but that may change in
+          the future.
+        """
+        pass
+
+    @abstractmethod
+    def set_model(self, model):
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks.py:96]: Do we need to make this one an `abstractmethod`? I do not think a user should override this?

File: python/orca/src/bigdl/orca/learn/pytorch/training_operator.py [hunk 11]
```diff
@@ -125,6 +125,7 @@ def __init__(self,
                         "multi-model or multi-optimizer training/validation.")
         self.timers = TimerCollection()
         self.setup(config)
+        self.batch_metric = None
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/training_operator.py:128]: where does this `batch_metric` is used?

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 12]
```diff
@@ -251,9 +252,15 @@ def train_epochs(self, data_creator, epochs=1, batch_size=32, profile=False,
                 loader = self.with_sampler(loader)
         elif wrap_dataloader is True:
             loader = self.with_sampler(loader)
+
+        for callback in callbacks:
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:256]: `callbacks` could be a None

Comment (yangw1234) [reply python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:256]:
```python
if callbacks is None:
    callbacks = []
```

**Comment by charlieJ107** [comment]
Manually tested on cpx-004. The example logging callback log the information as followed: 

> (pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_train_begin called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_epoch_begin called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:epoch got here: 0
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:On_batch_begin called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here 0
...
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_batch_end called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here: 0
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:On_batch_begin called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here 1
...
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_batch_end called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here: 1
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:On_batch_begin called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here 2
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_batch_end called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here: 2
...
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:On_batch_begin called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here 187
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_batch_end called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:batch got here: 187
(pid=64815) INFO:bigdl.orca.learn.pytorch.torch_runner:Finished training epoch 1, stats on rank 0: {'epoch': 1, 'batch_count': 188, 'num_samples': 60000, 'train_loss': 0.23345266155401867, 'last_train_loss': 0.1072411984205246}
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_epoch_end called
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:epoch got here: 1
(pid=64815) INFO:bigdl.orca.learn.pytorch.example_log_callback:on_train_end called
num_samples : 10000
Accuracy : tensor(0.9816)
val_loss : 0.06302541267871857
Stopping orca context

It seems it has passed the test.

**Comment by charlieJ107** [comment]
Jenkins passed at [BigDL-Orca-PR-Validation #238 [Jenkins]](http://10.112.231.51:18888/job/BigDL-Orca-PR-Validation/238/)

**Comment by yangw1234** [comment]
LGTM
